internal/log/coder.go

```diff
@@ -0,0 +1,136 @@
+package log
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"time"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/sloghuman"
+	"github.com/coder/coder/v2/agent/proto"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/codersdk/agentsdk"
+	"github.com/coder/retry"
+	"github.com/google/uuid"
+	"golang.org/x/mod/semver"
+)
+
+var (
+	rpcConnectTimeout  = 10 * time.Second
+	logSendGracePeriod = 10 * time.Second
+	minAgentAPIV2      = "v2.9"
+)
+
+// Coder establishes a connection to the Coder instance located at
+// coderURL and authenticates using token. It then establishes a
+// dRPC connection to the Agent API and begins sending logs.
+// If the version of Coder does not support the Agent API, it will
+// fall back to using the PatchLogs endpoint.
+// The returned function is used to block until all logs are sent.
+func Coder(ctx context.Context, coderURL *url.URL, token string) (Func, func(), error) {
+	client := initClient(coderURL, token)
+	bi, err := client.SDK.BuildInfo(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("get coder build version: %w", err)
+	}
+	if semver.Compare(semver.MajorMinor(bi.Version), minAgentAPIV2) < 0 {
+		sendLogs, flushLogs := sendLogsV1(ctx, client)
```

Comment:
Nit: I think we should log that we're using the old or the new API.

Reply:
fixed